streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -272,15 +280,22 @@ public synchronized void onChange(final StreamThread thread,
             }
             if (newState == StreamThread.State.PARTITIONS_REVOKED ||
                 newState == StreamThread.State.ASSIGNING_PARTITIONS) {
-                setState(State.REBALANCING);
+                setStateWhenNotShuttingDownOrNotRunning(State.REBALANCING);
             } else if (newState == StreamThread.State.RUNNING) {
+                // one thread is running, check others
                 for (final StreamThread.State state : threadState.values()) {
                     if (state != StreamThread.State.RUNNING) {
                         return;
                     }
                 }
-                setState(State.RUNNING);
+                setStateWhenNotShuttingDownOrNotRunning(State.RUNNING);
```

Comment:
What if, say we have 2 threads at the beginning, and then 1 thread shutdown due to an error, but the instance still keeps running with 1 thread left?

First of all, would the above ever happen? If yes, how should we set the state of the instance?